### ISSUE TITLE
[BUG] Single-GPU deployment and multi-GPU deployment have different results

### ISSUE THREAD

**Issue by feifan-sun** [huggingface/text-generation-inference#1678]
### System Info

ghcr.io/huggingface/text-generation-inference:1.4
platform: NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0
GPU: A100

### Information

- [X] Docker
- [ ] The CLI directly

### Tasks

- [X] An officially supported command
- [ ] My own modifications

### Single-GPU

```
model=HuggingFaceH4/zephyr-7b-beta
volume=$PWD/data
docker run --gpus='"device=0"' --shm-size 1g -p 8021:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:1.4 --model-id $model
```

### Multi-GPU
```
model=HuggingFaceH4/zephyr-7b-beta
volume=$PWD/data
docker run --gpus='"device=5,7"' --shm-size 1g -p 8021:80 -v $volume:/data ghcr.io/huggingface/text-generation-inference:1.4 --model-id $model
```
### Request parameters
```
{
    "inputs": "hello, how are you?",
    "parameters": {
        "max_new_tokens": 10,
        "do_sample": false,
        "decoder_input_details": true
    }
}
 ```

### Single-GPU response
```
{
    "generated_text": "\n\ni’m doing well, thanks for",
    "details": {
        "finish_reason": "length",
        "generated_tokens": 10,
        "seed": null,
        "prefill": [
            {
                "id": 1,
                "text": "<s>",
                "logprob": null
            },
            {
                "id": 6312,
                "text": "hell",
                "logprob": -12.6953125
            },
            {
                "id": 28709,
                "text": "o",
                "logprob": -0.052825928
            },
            {
                "id": 28725,
                "text": ",",
                "logprob": -1.4326172
            },
            {
                "id": 910,
                "text": "how",
                "logprob": -4.4648438
            },
            {
                "id": 460,
                "text": "are",
                "logprob": -1.2910156
            },
            {
                "id": 368,
                "text": "you",
                "logprob": -0.03933716
            },
            {
                "id": 28804,
                "text": "?",
                "logprob": -0.82421875
            }
        ],
        "tokens": [
            {
                "id": 13,
                "text": "\n",
                "logprob": -0.7133789,
                "special": false
            },
            {
                "id": 13,
                "text": "\n",
                "logprob": -0.54003906,
                "special": false
            },
            {
                "id": 28710,
                "text": "i",
                "logprob": -1.1416016,
                "special": false
            },
            {
                "id": 28809,
                "text": "’",
                "logprob": -1.4414062,
                "special": false
            },
            {
                "id": 28719,
                "text": "m",
                "logprob": -0.27685547,
                "special": false
            },
            {
                "id": 2548,
                "text": " doing",
                "logprob": -1.8203125,
                "special": false
            },
            {
                "id": 1162,
                "text": " well",
                "logprob": -1.1337891,
                "special": false
            },
            {
                "id": 28725,
                "text": ",",
                "logprob": -0.38745117,
                "special": false
            },
            {
                "id": 8196,
                "text": " thanks",
                "logprob": -0.9326172,
                "special": false
            },
            {
                "id": 354,
                "text": " for",
                "logprob": -0.47509766,
                "special": false
            }
        ]
    }
}
```

### Multi-GPU response

```
{
    "generated_text": "\n\ni’m doing well, thanks for",
    "details": {
        "finish_reason": "length",
        "generated_tokens": 10,
        "seed": null,
        "prefill": [
            {
                "id": 1,
                "text": "<s>",
                "logprob": null
            },
            {
                "id": 6312,
                "text": "hell",
                "logprob": -12.703125
            },
            {
                "id": 28709,
                "text": "o",
                "logprob": -0.05328369
            },
            {
                "id": 28725,
                "text": ",",
                "logprob": -1.4404297
            },
            {
                "id": 910,
                "text": "how",
                "logprob": -4.4570312
            },
            {
                "id": 460,
                "text": "are",
                "logprob": -1.2919922
            },
            {
                "id": 368,
                "text": "you",
                "logprob": -0.039367676
            },
            {
                "id": 28804,
                "text": "?",
                "logprob": -0.82714844
            }
        ],
        "tokens": [
            {
                "id": 13,
                "text": "\n",
                "logprob": -0.7138672,
                "special": false
            },
            {
                "id": 13,
                "text": "\n",
                "logprob": -0.54296875,
                "special": false
            },
            {
                "id": 28710,
                "text": "i",
                "logprob": -1.1425781,
                "special": false
            },
            {
                "id": 28809,
                "text": "’",
                "logprob": -1.4462891,
                "special": false
            },
            {
                "id": 28719,
                "text": "m",
                "logprob": -0.27856445,
                "special": false
            },
            {
                "id": 2548,
                "text": " doing",
                "logprob": -1.8164062,
                "special": false
            },
            {
                "id": 1162,
                "text": " well",
                "logprob": -1.1298828,
                "special": false
            },
            {
                "id": 28725,
                "text": ",",
                "logprob": -0.38891602,
                "special": false
            },
            {
                "id": 8196,
                "text": " thanks",
                "logprob": -0.93310547,
                "special": false
            },
            {
                "id": 354,
                "text": " for",
                "logprob": -0.48046875,
                "special": false
            }
        ]
    }
}
```

In this case, although the output is the same, the logprob is different from the beginning.

### Expected behavior

The logprob should be consistent.

**Comment by OlivierDehaene** [comment]
Yes that's expected. Nvidia CUDA kernels are not 100% deterministic and small approximations can accumulate.
When you shard on multiple GPUs, the selected kernels are not the same and the logits can be a bit different. 

As far as I know, there is no definitive answer in the litterature on wether this affect the downstream performance of the model.
